services/src/main/java/org/keycloak/protocol/oidc/mappers/ClaimsParameterWithValueIdTokenMapper.java

```diff
@@ -0,0 +1,149 @@
+import org.keycloak.models.ClientSessionContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ProtocolMapperModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.provider.ProviderConfigProperty;
+import org.keycloak.representations.IDToken;
+import org.keycloak.util.JsonSerialization;
+import org.keycloak.util.TokenUtil;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+/**
+ * @author <a href="mailto:takashi.norimatsu.ws@hitachi.com">Takashi Norimatsu</a>
+ */
+public class ClaimsParameterWithValueIdTokenMapper extends AbstractOIDCProtocolMapper implements OIDCIDTokenMapper {
+
+    public static final String PROVIDER_ID = "oidc-claims-param-value-idtoken-mapper";
+
+    private static final List<ProviderConfigProperty> configProperties = new ArrayList<>();
+
+    public static final String CLAIM_VALUE = "claim.value";
+
+    static {
+        ProviderConfigProperty property = new ProviderConfigProperty();
+        property.setName(CLAIM_VALUE);
+        property.setLabel("Claim value");
+        property.setType(ProviderConfigProperty.STRING_TYPE);
+        property.setHelpText("Value of the claim you want to set. 'true' and 'false can be used for boolean values.");
+        configProperties.add(property);
+
+        OIDCAttributeMapperHelper.addIncludeInTokensConfig(configProperties, ClaimsParameterWithValueIdTokenMapper.class);
+    }
+
+    @Override
+    public String getDisplayCategory() {
+        return TOKEN_MAPPER_CATEGORY;
+    }
+
+    @Override
+    public String getDisplayType() {
+        return "Claims parameter with value ID Token";
+    }
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public String getHelpText() {
+        return "Claims specified by Claims parameter with value are put into an ID token.";
+    }
+
+    @Override
+    public List<ProviderConfigProperty> getConfigProperties() {
+        return configProperties;
+    }
+
+    @Override
+    protected void setClaim(IDToken token, ProtocolMapperModel mappingModel, UserSessionModel userSession, KeycloakSession keycloakSession, ClientSessionContext clientSessionCtx) {
+        String claims = clientSessionCtx.getClientSession().getNote(OIDCLoginProtocol.CLAIMS_PARAM);
+        if (claims == null) return;
+
+        if (TokenUtil.TOKEN_TYPE_ID.equals(token.getType())) {
+            putClaims(OAuth2Constants.ID_TOKEN, claims, token, mappingModel, userSession);
+        }
+    }
+
+    private void putClaims(String tokenType, String claims, IDToken token, ProtocolMapperModel mappingModel, UserSessionModel userSession) {
+        String attributeValue = mappingModel.getConfig().get(CLAIM_VALUE);
+        if (attributeValue == null) return;
+
+        JsonNode requestParams = null;
+
+        try {
+            requestParams = JsonSerialization.readValue(claims, JsonNode.class);
```

Comment:
We have class `ClaimsRepresentation` . It provides some help for parsing claims parameter. Maybe it can be used to simplify a code below and avoid so much manual "JSON parsing" ?

Reply:
Yes, I will use `ClaimsRepresentation `.